### PR TITLE
v0.12 fork choice update

### LINKED ISSUE #1054: Ensure fork choice is to-spec
## Description

Fork choice needs to be checked to ensure it's up-to-spec. I suspect we're off on the following things:

- Justified checkpoint updating.
- Delaying consideration of attestations until they're at least 1 slot old.

I will come back to this one, just leaving some notes so I don't forget.

### LINKED ISSUE #1059: Follow up with redundant fork choice check
## Description

Over in https://github.com/ethereum/eth2.0-specs/pull/1755 I've made a suggestion to the spec that a certain check is redundant. I've already skipped the check in #983:

https://github.com/sigp/lighthouse/blob/8ea5397efa3718a43db1a0f7f6b3b4f77966f039/beacon_node/beacon_chain/src/attestation_verification.rs#L529-L552

We should keep an eye on this PR and resolve the TODO.

### LINKED ISSUE #1060: Follow up with FFG source/target check
## Description

Over in https://github.com/ethereum/eth2.0-specs/issues/1636 I've raised an issue about checking FFG source/target in fork choice attn processing. Here's our corresponding code in #983:

https://github.com/sigp/lighthouse/blob/8ea5397efa3718a43db1a0f7f6b3b4f77966f039/beacon_node/beacon_chain/src/attestation_verification.rs#L554-L561

We should keep an eye on this PR and resolve the TODO.

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

- Resolves #1054
- Resolves #1059
- Resolves #1060

## Proposed Changes

- Update fork choice to `v0.12.0`
- Refactors everything above `proto_array` to read more like the spec.
- Moves the `beacon_chain::ForkChoice` struct into a new `consensus/fork_choice` crate (along with a significant refactor).
- Renames the `proto_array_fork_choice` crate to `proto_array` (it seems less confusing now we have a `fork_choice` crate too).
- Removes `VerifiedForkChoiceAttestation` and shifts that logic into the new `fork_choice` crate.
  - Replaces `IntoForkChoiceVerifiedAttestation` with `SignatureVerifiedAttestation`.
- Moves `beacon_chain/tests/attestation_verification.rs` tests related to fork choice into the new `fork_choice` crate
  - Also refactors them into @michaelsproul's delicious builder-y-testing-struct pattern.
- Lifts all concurrency locks from `ProtoArray` and `ForkChoice` up to a single, monolithic lock on `beacon_chain.fork_choice`.
  - This makes it easier to reason about the consistency of fork choice at the risk of blocking some legit concurrent operations.
- Hold a write-lock on `beacon_chain.fork_choice` until an block has been added to fork choice *and* leveldb.
  - Once again, favours consistency at the risk of blocking some legit concurrent ops.
- Adds some new methods to `BeaconChainHarness` (nothing breaking).
- Removes the `beacon_chain::Builder::reduced_tree_fork_choice` method:
  - It was incorrectly named.
  - We *must* call it, so I just moved its functionality into `Builder::build`

## TODO

- [x] Delay consideration of attestations from this slot.
- [x] New verification conditions
- [x] ~~Harmony fork choice tests~~ (deferred to later PR)
- [x] ~~Fix this todo from #1228:~~ https://github.com/sigp/lighthouse/blob/f498081941b0b26e5cec13783cc3f3531cdd3e37/beacon_node/beacon_chain/tests/attestation_verification.rs#L123-L125
- [x] Fix double clone in best justified checkpoint
- [x] Rethink caching justified state balances
- [x] Add metrics again

### CODE REVIEW

**Comment by michaelsproul** [comment]
Do you mind taking the `TODO`s related to fork choice in attestation verification in as part of this PR?

https://github.com/sigp/lighthouse/blob/cb26ddebb1b9ec0ba24da287a4ed83fc1b4d9153/beacon_node/beacon_chain/src/attestation_verification.rs#L554-L561

https://github.com/sigp/lighthouse/blob/cb26ddebb1b9ec0ba24da287a4ed83fc1b4d9153/beacon_node/beacon_chain/src/attestation_verification.rs#L529-L537

I think it would make most sense for the target verification to be added alongside these fork-choice changes, as the underlying fork choice API seems to be changing a bit (and we're getting `get_ancestor`).

**Comment by michaelsproul** [comment]
> Fix this todo from #1228:

I already fixed this in #1236, so I'm ticking it off (it's not fork choice related)

**Comment by paulhauner** [comment]
This is ready for review, assuming it passes CI :)

File: consensus/fork_choice/src/fork_choice.rs [hunk 1]
```diff
@@ -0,0 +1,892 @@
...
+fn compute_start_slot_at_epoch<E: EthSpec>(epoch: Epoch) -> Slot {
+    epoch.start_slot(E::slots_per_epoch())
+}
+
+/// Called whenever the current time increases.
+///
+/// ## Notes
+///
+/// This function should only ever be passed a `time` that is less than, equal to or one greater
+/// than the previously passed value. I.e., it must be called each time the slot changes.
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply consensus/fork_choice/src/fork_choice.rs:129]: Yep that's right. Will remove this note.

File: consensus/fork_choice/src/fork_choice.rs [hunk 2]
```diff
@@ -0,0 +1,892 @@
...
+    }
+
+    /// Returns the block root of an ancestor of `block_root` at the given `slot`. (Note: `slot` refers
+    /// to the block is *returned*, not the one that is supplied.)
+    ///
+    /// If `state_opt.is_some()`, the lookup will be much more efficient. However, the
+    /// state must fulfil the following conditions:
+    ///
+    /// 1. It must be from `block_root` or an **descendant** of `block_root`.
+    /// 2. It must have a equal to or greater than `ancestor_slot`.
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply consensus/fork_choice/src/fork_choice.rs]: Turns out that section of comment was outdated, so I removed it :)

File: consensus/fork_choice/src/fork_choice.rs [hunk 3]
```diff
@@ -0,0 +1,892 @@
...
+        }
+
+        if target.epoch != indexed_attestation.data.slot.epoch(E::slots_per_epoch()) {
+            return Err(InvalidAttestation::BadTargetEpoch {
+                target: target.epoch,
+                slot: indexed_attestation.data.slot,
+            });
+        }
+
+        // Attestation target must be for a known block.
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply consensus/fork_choice/src/fork_choice.rs:594]: I added a note :)

**Comment by paulhauner** [comment]
All comments addressed! There are two interesting commits since your review:

- https://github.com/sigp/lighthouse/pull/1229/commits/17c8fa51a74c8e55f8708bab48f91a651666104c: Makes some changes to `get_ancestor`:
  - Fixes a bug with skip slots
  - Returns an `Option` so we can gracefully detect when we've iterated outside of proto array.
- https://github.com/sigp/lighthouse/pull/1229/commits/1b6ccb05979461015996f1d7d4cee5d9407f34a3: Addresses comments from @michaelsproul.

File: beacon_node/beacon_chain/tests/attestation_verification.rs [hunk 4]
```diff
@@ -87,7 +87,7 @@ fn get_valid_unaggregated_attestation<T: BeaconChainTypes>(
 
 fn get_valid_aggregated_attestation<T: BeaconChainTypes>(
     chain: &BeaconChain<T>,
-    aggregate: Attestation<T::EthSpec>,
+    mut aggregate: Attestation<T::EthSpec>,
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/tests/attestation_verification.rs:90]: These were part of the temporary FIXME change, and will trigger warnings now, so should be reverted (I like `git checkout -p` for this)

File: beacon_node/beacon_chain/tests/attestation_verification.rs [hunk 5]
```diff
@@ -97,10 +97,11 @@ fn get_valid_aggregated_attestation<T: BeaconChainTypes>(
         .expect("should get committees");
     let committee_len = committee.committee.len();
 
-    let (aggregator_index, aggregator_sk) = committee
+    let (aggregator_committee_pos, aggregator_index, aggregator_sk) = committee
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/tests/attestation_verification.rs:100]: Can delete

File: beacon_node/beacon_chain/tests/attestation_verification.rs [hunk 6]
```diff
@@ -97,10 +97,11 @@ fn get_valid_aggregated_attestation<T: BeaconChainTypes>(
...
         .committee
         .iter()
-        .find_map(|&val_index| {
+        .enumerate()
+        .find_map(|(committee_pos, &val_index)| {
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/tests/attestation_verification.rs:104]: Can revert

File: beacon_node/beacon_chain/tests/attestation_verification.rs [hunk 7]
```diff
@@ -112,7 +113,7 @@ fn get_valid_aggregated_attestation<T: BeaconChainTypes>(
             );
 
             if proof.is_aggregator(committee_len, &chain.spec).unwrap() {
-                Some((val_index, aggregator_sk))
+                Some((committee_pos, val_index, aggregator_sk))
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/tests/attestation_verification.rs:116]: Not reqd

**Comment by paulhauner** [comment]
This one should revert that old FIXME change :)

https://github.com/sigp/lighthouse/pull/1229/commits/00bb20a42378f7a3ef71ad9d166e020e3ec02c43

**Pull request merged**